### PR TITLE
enable compiled Liger JSD losses

### PR DESCRIPTION
**Pull request by mdeff**
# What does this PR do?

This PR enables `torch.compile` for `LigerFusedLinearJSDLoss` by removing TRL's explicit `compiled=False` override.

It updates all six trainers using Liger JSD:
- `DistillationTrainer`
- `GKDTrainer`
- `GOLDTrainer`
- `IWOPDTrainer`
- `SDFTTrainer`
- `SDPOTrainer`

The corresponding DistillationTrainer parity test now also exercises Liger's compiled default.

## Motivation

`LigerFusedLinearJSDLoss` defaults to `compiled=True`, and Liger's correctness tests exercise that default. Other Liger chunked losses used by TRL, including DPO and KTO, already use compilation.

TRL has explicitly disabled compilation for JSD since its initial integration in #3946. I found no documented correctness or compatibility reason for this override in the original PR or subsequent JSD fixes and refactors, including #5731 and #6537.

## Benchmark

The benchmark used the same single H100 environment for all configurations:
- Llama 3.2 1B student and teacher
- bfloat16
- sequence length 1024
- batch size 32
- 1,000 optimizer steps
- fused AdamW
- no gradient checkpointing
- no evaluation or checkpoint saving

Elapsed time was measured from the start of training through completion, so the initial compilation cost is included.

| Chunk size | Uncompiled | Compiled | Runtime reduction | Throughput gain |
|---:|---:|---:|---:|---:|
| 512 | 1.224 s/step, 26.8k tok/s | 0.874 s/step, 37.5k tok/s | 28.6% | 40.1% |
| 1024 | 1.202 s/step, 27.3k tok/s | 0.832 s/step, 39.4k tok/s | 30.8% | 44.6% |
| 2048 | 1.165 s/step, 28.1k tok/s | 0.829 s/step, 39.5k tok/s | 28.8% | 40.5% |
| 4096 | 1.157 s/step, 28.3k tok/s | 0.811 s/step, 40.4k tok/s | 29.9% | 42.7% |

Across the four tested chunk sizes, compilation reduced elapsed training time by approximately 29.5% and increased throughput by approximately 42%.

## Tests

The affected Liger tests were run with TRL's CI configuration:
- PyTorch 2.8.0+cu128
- Liger Kernel 0.8.2
- fresh TorchInductor cache

Result:
```text
14 passed, 1 skipped, 257 deselected
```

The suite covers the compiled JSD paths in Distillation, GKD, GOLD, IW-OPD, SDFT, and SDPO, including loss-parity and end-to-end training tests.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

---

> [!NOTE]
> **Medium Risk**
> Training loss now goes through `torch.compile` on the fused JSD path, which can affect numerics, warmup/compile behavior, and PyTorch/Liger version compatibility despite existing parity tests.
> 
> **Overview**
> Stops forcing `compiled=False` on `LigerFusedLinearJSDLoss` so TRL’s distillation trainers use Liger’s default **torch.compile** path for fused JSD.
> 
> The change is applied everywhere TRL wires up that loss: **`DistillationTrainer`**, **`GKDTrainer`**, **`GOLDTrainer`**, **`IWOPDTrainer`**, **`SDFTTrainer`**, and **`SDPOTrainer`**. **`test_liger_loss_agrees_with_chunked`** in `test_distillation_trainer.py` is updated the same way, so parity checks run against the compiled kernel instead of an explicitly uncompiled instance.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 720bc4634f8ebd3f13bee691e45af359312fbeb0. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by behroozazarkhalili** [comment]
I checked the four factual claims behind this change. All four hold, and the consistency argument is stronger than the PR body makes it.

**The diff does what it says.** Constructing the loss with exactly the kwargs the trainers now use, on liger-kernel 0.8.2:

```
compiled default = True
constructed OK, compiled = True
```

So removing the override really does hand the path to Liger's default rather than relying on anything implicit.

**Completeness.** There are six `LigerFusedLinearJSDLoss` construction sites under `trl/` plus the one in `tests/test_distillation_trainer.py`, and the PR touches all seven. After it, no Liger loss construction anywhere in `trl/` passes `compiled` at all: the DPO, KTO, and GRPO sites never did. That makes JSD uniform with every other fused loss in the repo, which is the argument I would lead with, since TRL treats consistency across duplicated call sites as a requirement rather than a preference.

**Origin of the override.** `39cc9a82` (`[GKD] add liger loss`, #3946) introduced `compiled=False` with no comment and no rationale in the diff, and nothing since added one. Your reading of the history matches mine.

**Gates.** ruff 0.13.3 `check` and `format --check` and the pinned doc-builder are all clean on all seven files.

Two things to flag, neither of them a blocker.

**`pyproject.toml:253` becomes inaccurate.** That filter carries the comment:

```
# Triggered by the DPO and KTO Liger tests, the only ones that compile, since the GPU CI moved from T4 to L40S
```

After this PR the JSD tests compile too, so "the only ones that compile" is no longer true. Nothing breaks: the entry is an `ignore` and the `filterwarnings` list has zero `error` entries, so CI cannot fail on that advisory either way. The cost is a comment that now misdescribes the situation, on a filter linked to tracking issue #6849. A one-line edit alongside this change would keep it honest.

**What the green parity test does and does not establish.** `test_liger_loss_agrees_with_chunked` carries its own caveat: it scales both LM heads by 50 because, at the float32 noise floor with a near-uniform distribution over the ~150k vocab, the chunked and fused paths' matmul reduction orders diverge by about 2x. So the assertion at `rtol=1e-3, atol=1e-4` is parity on a deliberately well-conditioned batch. That is a reasonable test design, and it is also narrower than "compilation leaves the numerics alone", which is where Bugbot's medium-risk note points. Worth saying out loud in the PR body so a later reader does not over-read the green check.

One limitation on my side, stated plainly: I could not reproduce your benchmark or run the Liger tests. This machine has no GPU (`torch.cuda.is_available()` is `False`, device count 0) and the cluster queue is long enough that a run is not practical right now. Everything above is source-level and CPU-level verification. The 29.5% runtime and 42% throughput figures and the compiled-path test results are yours, not independently confirmed by me.

On the substance I think this should go in. A 29% runtime reduction for deleting seven lines is a good trade, the override had no recorded reason, and the change moves JSD toward the rest of the repo instead of away from it.

**Comment by kashif** [comment]
Heads up: #7301 removes the fused Liger JSD loss from these five trainers entirely (matching #7064 for `DistillationTrainer`), as part of #7063's plan to drop the Liger fused-loss paths in favor of the chunked log-prob path everywhere. Once that lands, this PR's `compiled=False` removals won't have anything left to apply to. Sorry for the wasted effort — thanks for digging into the compile flag.

**Comment by behroozazarkhalili** [comment]
Thanks for the heads-up. Agreed, once #7301 lands there is nothing left here for the compile flag to apply to, and dropping the fused path is the cleaner outcome anyway. It also settles the `pyproject.toml` filterwarnings note I raised above: with the JSD path gone, the DPO and KTO Liger tests really are the only ones that compile, so that comment stays accurate as written.
